src/poetry/packages/locker.py

```diff
@@ -129,8 +129,23 @@ def locked_repository(self) -> Repository:
                 hashes = cast("dict[str, Any]", metadata["hashes"])
                 package.files = [{"name": h, "hash": h} for h in hashes[name]]
             else:
-                files = metadata["files"][name]
-                package.files = files
+                if source_type in {"git", "directory", "url"}:
+                    package.files = []
+                else:
+                    files = metadata["files"][name]
+                    if source_type == "file":
+                        if len(files) > 1:
+                            filename = Path(url).name
+                            for file_item in files:
+                                if file_item["file"] == filename:
+                                    package.files = [file_item]
+                                    break
+                    else:
+                        # Strictly speaking, this is not correct, but we have no chance
+                        # to always determine which are the correct files.
+                        # This is fixed later in Provider.complete_package() by looking
```

Comment:
this is only true in code paths that perform solving.  Eg in `poetry install` there is no lookup in the repository, right?

Reply:
`poetry install` performs a solver run, too.
https://github.com/python-poetry/poetry/blob/5997a605e55c9cbf08e70d60cd1980e45d94574b/src/poetry/installation/installer.py#L312-L316

Reply:
I'm pretty sure it doesn't reach complete_package and update the `files` on things that it read from the locker.

If it did, that would defeat the point of storing the hashes in the first place.  The whole purpose is to check that the hashes for the files that are installed are the same as the hashes in the lockfile: if we always refreshed those values before making the check then this would all be a big old waste of time.

Reply:
Of course, you're right. `complete_package()` is called but it does not do anything because in this solver run the only repository in the pool is the repository created from the lockfile. I'll adapt the comment.

Reply:
Suggest that it's sufficient to comment this code, without reference to other code.  ie what you're saying in your first sentence is enough: this lockfile doesn't accurately record which files go with which packages, and this is our best effort at reconstructing that.

Other parts of the codebase can look after themselves!